reference-implementation/unittests/mhlo.cpp

```diff
@@ -31,160 +31,393 @@ const float EPSILON = 5e-4;
 TEST(mhlo, abs) {
   EXPECT_EQ(1, mhlo::abs(-1));
 
-  Tensor0D<int> t0{-1};
-  Tensor1D<float, 2> t1{-1.0f, -2.0f};
-  Tensor2D<long, 2, 2> t2{-2, -1, 0, 2};
-  Tensor3D<int32_t, 2, 1, 2> t3{-2, -1, 0, 2};
-  Tensor4D<int, 2, 2, 1, 2> t4{-2, -1, 0, 0, 3, -3, -2, 1};
-
-  EXPECT_THAT(mhlo::abs(t0), Pointwise(Eq(), {1}));
-  EXPECT_THAT(mhlo::abs(t1), Pointwise(Eq(), {1.0f, 2.0f}));
-  EXPECT_THAT(mhlo::abs(t2), Pointwise(Eq(), {2, 1, 0, 2}));
-  EXPECT_THAT(mhlo::abs(t3), Pointwise(Eq(), {2, 1, 0, 2}));
-  EXPECT_THAT(mhlo::abs(t4), Pointwise(Eq(), {2, 1, 0, 0, 3, 3, 2, 1}));
+  {
+    Tensor0D<int> x{-1};
+    Tensor0D<int> expected_result{1};
+    Tensor0D<int> result = mhlo::abs(x);
```

Comment:
Should `Tensor0D` be the "lowest level" or should we also add tests for the fundamental datatype, similar to what we do in `arith.cpp`:
https://github.com/iml130/mlir-emitc/blob/d180553cc82b42f460ca8aea57491fc9d18c8fa2/reference-implementation/unittests/arith.cpp#L26-L39

Reply:
We should check against the op [definitions](https://github.com/tensorflow/mlir-hlo/blob/402eeff6db40e3790eac6ff82b14fef67b86c0d7/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td#L152) if scalar types are supported and change the [signature](https://github.com/iml130/mlir-emitc/blob/d180553cc82b42f460ca8aea57491fc9d18c8fa2/reference-implementation/include/emitc/mhlo.h#L603) if necessary.

In this case the abs function should be deleted for scalar types via a `IsTensor<Src> = true` template argument.

Reply:
In that case, I would suggest to refactor our kernels in a follow up.

Reply:
@heidebro Please fill an issue in our internal tracker.